4317 ai auto update debounce

cezary-baserow opened the pull request:
depends on https://github.com/baserow/baserow/pull/4263

fixes #4317 

### What is in this PR

Adds debouncing for AI field value generation, if the generation was triggered from auto-update.

If an AI field refers to another field, and auto-update is enabled, changes on values for that another field will trigger tasks for each set of rows updated. There can be many small changes in a short period of time, which may generate many update tasks for the AI field (including many updates to the same row). 

This fix will add a grace period between scheduling those AI field value updates and generating AI field values. Rows that are scheduled for generation will be collected in a separate table (`AIFieldScheduledUpdate` model). Within one grace period, rows will be de-duplicated, so one row will be updated only once for a specific field.

The task that generates the values will retry, if there are rows in the scheduling table, that haven't been processed yet.

### How to test this PR
- Use a table with AI field, where prompt will refer another field.
- ensure auto-update is enabled on that field
- modify referred field in several rows in a short period of time
- observe rows being populated.
- if you modify more rows during ai value generation, this should also update them once initial batch has been processed.

### Checklist

- [x] A changelog entry has been added to `changelog/entries/unreleased` using `changelog/src/changelog.py`
- [x] New/updated **Premium/Enterprise features** are separated correctly in the premium or enterprise folder
- [ ] The latest **Chrome and Firefox** have been used to test any new frontend features
- [x] [Documentation](https://github.com/baserow/baserow/blob/master/docs/index.md) has been updated
- [ ] [Quality Standards](https://github.com/baserow/baserow/blob/master/CONTRIBUTING.md#quality-standards) are met
- [ ] **Performance**: tables are still fast with 100k+ rows, 100+ field tables
- [ ] The [redoc API pages](https://api.baserow.io/api/redoc/) have been updated for any REST API changes
- [ ] Our [custom API docs](https://github.com/baserow/baserow/blob/master/web-frontend/modules/database/pages/APIDocsDatabase.vue) are updated for changes to endpoints accessed via API tokens
- [ ] The UI/UX has been updated following the [UI Style Guide](https://baserow.io/style-guide)
- [ ] Security impact of change has been considered

dimmur-brw commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,163 @@
+from baserow.core.jobs.handler import JobHandler
+
+PERIODIC_CHECK_MINUTES = 5
+PERIODIC_CHECK_TIME_LIMIT = 60 * PERIODIC_CHECK_MINUTES  # 15 minutes.
```
comment says 15 minutes, but variable above is set to 5

dimmur-brw commented on premium/backend/src/baserow_premium/fields/job_types.py:
```diff
@@ -612,17 +630,17 @@ def handle_result(self, row: GeneratedTableModel, result: Exception | Any):
                 self.update_value(row, result)
         finally:
-            self.update_progress(row)
+            self.update_progress(row, result)
```
It might be this place that return `Exception` as result. Would it make sense to handle it here instead of passing down?

dimmur-brw reviewed (changes requested):
This is pretty cool feature 💯 

I found few minor things, posted in comments

cezary-baserow commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,163 @@
+
+    try:
+        _schedule_generate_ai_value_generation(field_id, retry=retry)
+    except DuplicateTaskError:
```
`DuplicateTaskError` name is pretty self-explainatory

dimmur-brw reviewed (approved):
Looks good 🚀 
Let's merge after CI is green

silvestrid commented on premium/backend/src/baserow_premium/fields/job_types.py:
```diff
@@ -275,7 +285,30 @@ def run(self, job: GenerateAIValuesJob, progress):
+            rows_progress.increment()
+
+            if job.is_auto_update and not isinstance(value, Exception):
+                AIFieldScheduledUpdate.objects.filter(
```
what happens if the the cell values changed during the AI generation. This query will delete a pending update and that update will be lost.

This query should update a timestamp with the time when the value has been generated instead, letting some other code to verify if the entry can be deleted or needs to be generated again.

silvestrid commented on premium/backend/src/baserow_premium/fields/models.py:
```diff
@@ -93,12 +95,59 @@ class MODES(StrEnum):
+        null=False, help_text="The ID of the field to update."
+    )
+    row_id = models.IntegerField(null=False, help_text="Row ID to update")
+    updated_on = models.DateTimeField(
```
I think we need 2 datetimes here:
- one datetime to indicate when an update has been scheduled
- one datetime to indicate when a new value is generated

Only if generated_at > scheduled_at we can delete rows from this table

silvestrid replied:
Yes, It can be done with only one as well, as long as we have the datetime when the data has been read from the db. It doesn't have to check if the timestamp has changed in the meantime, it would be enough to filter properly. Something like: `AIFieldScheduledUpdate.objects.filter(field_id=ai_field.id, row_id=row.id, updated_on__lte=ref_fields_read_at).delete()` or maybe even better `_raw_delete()` since delete can be slow

silvestrid commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,176 @@
+    soft_time_limit=settings.CELERY_SEARCH_UPDATE_HARD_TIME_LIMIT,
+    time_limit=settings.CELERY_SEARCH_UPDATE_HARD_TIME_LIMIT,
+)
+def generate_scheduled_ai_field_generation(field_id: int, retry: int = 3):
```
[minor] is there a reason to implement our own retry logic when `celery` has a builtin feature for it?

silvestrid replied:
Also, I'm not sure I fully understand how this retry works and why it's needed. It looks to me that the job won't be able to reschedule itself more than 3 times, even if there are stil pending updates. Why?
Wouldn't the periodic reschedule start another job anyway? It feels like unnecessary logic, unless I'm missing something

silvestrid commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,176 @@
+    """
+
+    return list(
+        AIFieldScheduledUpdate.objects.filter(
```
Shouldn't we limit this list? If we have 1M rows, the task will almost certainly fail. Wouldn't be better to submit only a batch of, let's say, 100-200 rows at a time?

cezary-baserow replied:
Yep, this can be done in batches. Will change it, so each task will process up to 100 rows, then reschedule next run, if there's more rows to process. Periodic task may also kick in, but it's ok, since it's purpose is to process any leftover rows.

cezary-baserow replied:
Actually, this will have a side-effect in the UI: auto-update jobs will appear in the list of AI field jobs. But this will also happen with a scheduled cleanup taks, which will schedule a job, if some rows are still waiting.

silvestrid commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,176 @@
+        "field_id"
+    ):
+        # we can send an empty list. This will reschedule all pending rows.
+        schedule_ai_field_generation(field_id=field_id, row_ids=[])
```
[minor] Instead of wriging a comment to explain why is allow to schedule an empty list of rows, which is a bit odd, wouldn't be simpler/clearer to have the try/schedule/except Duplicate error here instead?

silvestrid commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,176 @@
+    )
+
+    AIFieldScheduledUpdate.objects.filter(updated_on__lte=cutoff).delete()
+    for [field_id] in AIFieldScheduledUpdate.objects.distinct("field_id").values_list(
```
[minor] wdyt of using `values_list(flat=True)` instead of `for [field_id] in `?

silvestrid reviewed (changes requested):
Hey Cezary, the changes work, although I have some feedback on how concurrent updates and job's errors are handled. I'd love to hear you thoughts

cezary-baserow commented on premium/backend/src/baserow_premium/fields/models.py:
```diff
@@ -93,12 +95,59 @@ class MODES(StrEnum):
+    )
+
+    class Meta:
+        constraints = [
```
`unique_together` is not recommended by the docs: https://docs.djangoproject.com/en/5.0/ref/models/options/#django.db.models.Options.unique_together

silvestrid replied:
Ah, good to know! Let's keep it then

cezary-baserow commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,176 @@
+
+    user = ai_field.ai_auto_update_user
+
+    if row_ids := get_scheduled_ai_field_updates(field_id, now):
```
Yep, it's not needed.

cezary-baserow commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,176 @@
+                sync=True,
+            )
+        except Exception as e:
+            logger.error(f"Job failed: {e}")
```
I'm not sure if this works here in this way. This job is executed within a task, so the task will be killed, not the job. However, a job can be cancelled from UI. Should we remove rows that were scheduled for that job (so a batch, not all rows)?

silvestrid commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,176 @@
+    jh = JobHandler()
+    ai_field = AIField.objects.select_related("ai_auto_update_user").get(id=field_id)
+
+    user = ai_field.ai_auto_update_user
```
yes. Just to confirm, the next time the auto-update runs it will check that the user is None and disable the auto-update, right?

cezary-baserow replied:
Updated: if there's no auto-update user, the task removes pending row ids for the field and exits early.

cezary-baserow commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,176 @@
+        hours=settings.HOURS_UNTIL_TRASH_PERMANENTLY_DELETED
+    )
+
+    AIFieldScheduledUpdate.objects.filter(updated_on__lte=cutoff).delete()
```
removed `until` usage.

cezary-baserow commented:
@silvestrid I've updated the PR:
* removed `until` usage
* if there's a problem with AI field in the task, scheduled row ids are removed for the field, and no job is executed
* in the job handling, a result can be discarded for a row that has been executed from auto-update and the job couldn't remove a scheduled row id with `updated_on` value before the row started being processed. 

It's not tested well though.

dimmur-brw commented on premium/backend/src/baserow_premium/fields/job_types.py:
```diff
@@ -588,48 +634,35 @@ def is_finished(self) -> bool:
         """
 
         row = next(rows_iter)
+        if row.trashed:  # Skip trashed rows
```
we can remove comment

silvestrid replied:
Actually, there's no reason to do this check here. The query to fetch rows should not return trashed rows, and if the row is trashed during the generation, this check won't prevent `update_row_by_id` from failing to update the row, so I removed it

silvestrid commented on premium/backend/src/baserow_premium/fields/models.py:
```diff
@@ -93,12 +90,55 @@ class MODES(StrEnum):
+        help_text="If set, the job has been scheduled as a result of AI field auto-update.",
+    )
+
+    # TODO: no longer needed. Remove in a feature release
```
No, but IMO TODOs in the codebase should work better than issues in this case, because it's easy to forget about existing non-prioritized issues, while TODOs in the code are always there

silvestrid commented on premium/backend/src/baserow_premium/fields/tasks.py:
```diff
@@ -0,0 +1,173 @@
+        GenerateAIValuesJobType.type,
+        field_id=field_id,
+        is_auto_update=True,
+        sync=True,
```
It's crucial to run it with sync=True for everything to work correctly, because we want this to be a singleton and to have only one job of this type per field_id.

> What if job takes long time?

Like every other job, it fails with a timeout, but this one also release the lock so another job of the same type for the same field can start.

silvestrid commented on premium/backend/src/baserow_premium/fields/job_types.py:
```diff
@@ -275,7 +302,65 @@ def run(self, job: GenerateAIValuesJob, progress):
+                    values_already_prepared=True,
+                )
+            except Exception:
+                logger.exception(
```
It was added on purpose by Cezary to avoid stopping the generation in the middle, but it's not wise to continue and just log error in Sentry. This is a valid reason to stop the generation as well.

I updated the code to handle it properly and to return some more information about the exception that stopped the job. You can verify it by adding an invalid model and try to generate a value with it. For example with ollama now I can see: `AI model responded with errors: model 'invalid' not found (status code: 404)` in the database.

It's not exposed to the UI yet, but we can do that in a separate PR to provide more feedback to the user about what went wrong.
